google/api_core/retry_streaming_async.py

```diff
@@ -0,0 +1,260 @@
+                "Timeout of {:.1f}s exceeded".format(self.timeout),
+                source_exception,
+            ) from source_exception
+
+    async def _ensure_active_target(self) -> AsyncIterator[T]:
+        """
+        Ensure that the active target is populated and ready to be iterated over.
+
+        Returns:
+          - The active_target iterable
+        """
+        if not self.active_target:
+            new_iterable = self.target_fn()
+            if isinstance(new_iterable, Awaitable):
+                new_iterable = await new_iterable
+            self.active_target = new_iterable.__aiter__()
+        return self.active_target
+
+    def __aiter__(self) -> AsyncIterator[T]:
+        """Implement the async iterator protocol."""
+        return self
+
+    async def _handle_exception(self, exc) -> None:
+        """
+        When an exception is raised while iterating over the active_target,
+        check if it is retryable. If so, create a new active_target and
+        continue iterating. If not, raise the exception.
+        """
+        if not self.predicate(exc) and not isinstance(exc, asyncio.TimeoutError):
+            raise exc
+        else:
+            # run on_error callback if provided
+            if self.on_error:
+                self.on_error(exc)
+            try:
+                next_sleep = next(self.sleep_generator)
+            except StopIteration:
+                raise ValueError("Sleep generator stopped yielding sleep values")
+            # if deadline is exceeded, raise RetryError
+            if self.deadline is not None:
+                next_attempt = datetime_helpers.utcnow() + datetime.timedelta(
+                    seconds=next_sleep
+                )
+                self._check_timeout(next_attempt, exc)
+            _LOGGER.debug(
+                "Retrying due to {}, sleeping {:.1f}s ...".format(exc, next_sleep)
+            )
+            # sleep before retrying
+            await asyncio.sleep(next_sleep)
+            self.active_target = None
+            await self._ensure_active_target()
+
+    async def _iteration_helper(self, iteration_routine: Awaitable) -> T:
+        """
+        Helper function for sharing logic between __anext__ and asend.
+
+        Args:
+          - iteration_routine: The coroutine to await to get the next value
+                from the iterator (e.g. __anext__ or asend)
+        Returns:
+          - The next value from the active_target iterator.
+        """
+        # check for expired timeouts before attempting to iterate
+        self._check_timeout(datetime_helpers.utcnow())
+        try:
+            # grab the next value from the active_target
+            # Note: interrupting with asyncio.wait_for is expensive,
+            # so we only check  for timeouts at the start of each iteration
+            return await iteration_routine
+        except (Exception, asyncio.CancelledError) as exc:
+            await self._handle_exception(exc)
```

Comment:
Do we want to do anything special in `_handle_exception` if the exception is `CancelledError`?

Reply:
Good question... The default behavior is that cancellations will kill a generator. I added the extra CancelledError catch so that [cancellations could be retryable](https://github.com/googleapis/python-api-core/blob/14c195cac19606870e0d43c71a2c5409b7ba5811/tests/asyncio/test_retry_async.py#L546). (They usually aren't eligible for retries since they inherit from BaseException instead of Exception)

Looking at it again though, I'm not sure if this belongs in the wrapper. Users can always add a try/catch inside their wrapped generators if they want special handling for cancellations, and that probably makes more sense than handling it here. I think I'll remove this extra catch